[CUDA][Linalg] Add a `driver=` kwarg to `torch.linalg.svd` and `svdvals`; add cusolver gesvdaStridedBatched driver to svd

xwang233 opened the pull request:
[CUDA][Linalg] Add a driver= kwarg to torch.linalg.svd and svdvals; add cusolver gesvdaStridedBatched driver to svd

cusolver doc: https://docs.nvidia.com/cuda/cusolver/index.html#cuSolverDN-lt-t-gt-gesvda

Todo:

- [X] add cusolver `gesvdaStridedBatched` driver
- [X] add `driver=` kwarg to `torch.linalg.svd` and `torch.linalg.svdvals`
- [X] doc
- [X] error out (?) on other non-cusolver use cases: CPU, MAGMA
- [X] change svd api in `torch/csrc/api/include/torch/linalg.h` ?

Close https://github.com/pytorch/pytorch/issues/41306

Related https://github.com/pytorch/pytorch/issues/75494

xwang233 committed:
init

xwang233 committed:
change function name

xwang233 committed:
add gesvda dispatch

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
add svd test

xwang233 committed:
add gesvdaStridedBatched implementation

xwang233 committed:
add comments

xwang233 commented:
cc @Lezcano @IvanYashchuk , need help on test failures here https://github.com/pytorch/pytorch/runs/5638637024?check_suite_focus=true

There are several gradient tests for `pinv_singular` failed with the new `gesvda` driver. It seems like this driver doesn't work well with input matrices being singular or too ill-conditioned (>10^9). The `svd` output of `U` may contain `nan` in this case, and that leads to failures in analytical and numerical gradients.

Lezcano commented:
My suggestion was the following:
SVD algorithms are unstable for matrices with repeated singular values (or a small singular-value gap), like singular matrices with 2 zeros, or small perturbations of the identity. This is something that also happens with all the backends we currently have.

Before, we used to fall back when an algorithm encounters convergence issues to `gesvd`. Now, we see that for some matrices, this new algorithm converges (info == 0) but to an incorrect value (very large `residual_frobenius_norm`). @xwang233 reported that for these inputs, `gesvd` also fails, but it looks like `gesvdj` converges. I'd say that, if the backend heuristic we have proposes to go for `gesvda` and this one fails (note that this is very unlikely, but it may happen), we then default back to `gesvdj`, as we know that this one may converge, and if this one does not converge either, we go to `gesvd`.

My take on why doing this is that, even though these cases will be slower:
- It's better to be slow than to error out. Note that there's nothing weird with inputs with repeated singular values, `torch.eye(n)` is one such an input.
- Even if we have this fallback, it will barely ever be executed, so it's amortised cost is rather low.

IvanYashchuk commented:
> SVD algorithms are unstable for matrices with repeated singular values (or a small singular-value gap), like singular matrices with 2 zeros, or small perturbations of the identity. This is something that also happens with all the backends we currently have.

To make sure I understand this clearly. It's not SVD algorithms that are unstable for such matrices, but the derivative computation, right?

Lezcano commented:
The iterative algorithms I'm aware of for computing the SVD have problems dealing with matrices with small gap between their singular values. In fact, their accuracy / iterations needed (i.e. their complexity) often depends on this singular value gap.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
move _get_svd_outputs_sizes_and_strides outside of svd meta_func

xwang233 committed:
add `str? driver` optional argument to linalg_svd

xwang233 committed:
pass 'driver' arg to svd_stub

xwang233 committed:
add svd driver options

xwang233 committed:
add driver to torch.svd

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
gesvdj is a string

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
add driver to linalg.svdvals

xwang233 committed:
update svd heuristics

xwang233 committed:
add gesvdj tolerance and max sweeps

xwang233 committed:
change to a warning

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
revert meta linalg_out changes

xwang233 committed:
[to be reverted?] print better error message

Lezcano commented:
This PR is doing many things at the same time. Could you split it into a stack of PRs?

xwang233 committed:
tolerance

xwang233 committed:
Revert "[to be reverted?] print better error message"

This reverts commit 899ace695c65fda4f4ee8f3dcf3d5b07a9eca478.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
doc

xwang233 committed:
error out on non-cusolver backend

xwang233 committed:
change linalg.h function signature

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 commented:
Hi @Lezcano , I'm not familiar with the ghstack tool, and I think this PR does two coherent things which should be put together. The new `gesvda` driver needs to be exposed to make it meaningful. It can't be exposed without a kwarg driver (won't be used by default svd heuristics).

The diffs: 

- Cusolver bindings of `gesvdaStridedBatched` are in `cusolver.cpp` and `cusolver.h`
- Pytorch svd gesvda binding is in `BatchLinearAlgebraLib.cpp`
- Doc changes
- All other changes are mostly one-liner: add a `driver` argument to svd function signatures.

Please let me know if any part of the code is not clear. Thanks!

mruberry commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2992,6 +2994,10 @@ TORCH_IMPL_FUNC(_linalg_svd_out)(const Tensor& A,
   const bool use_cusolver = at::native::svd_uses_cusolver(A);
+  if (!use_cusolver && driver.has_value()) {
+    TORCH_CHECK(false,
+      "torch.linalg.svd: keyword argument `driver=` is only supported on CUDA inputs with cuSOLVER backend.");
```
Nice error message

mruberry commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -561,8 +568,8 @@ inline static void apply_svd_cusolver_gesvdjBatched(const Tensor& A, const Tenso
-  // TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetTolerance(gesvdj_params, 1.0e-7));
-  // TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetMaxSweeps(gesvdj_params, 15));
+  TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetTolerance(gesvdj_params, std::numeric_limits<scalar_t>::epsilon()));
+  TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetMaxSweeps(gesvdj_params, 400));
```
Nit: Might want to make the 400 a constexpr value so it can be changed consistently in one place in the future

mruberry commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -640,31 +718,72 @@ void svd_cusolver(const Tensor& A,
+      svd_cusolver_gesvdaStridedBatched(cloneBatchedColumnMajor(A), U, S, V, info, full_matrices, compute_uv);
+      convergence_check_needed = true;
+    } else {
+      TORCH_CHECK(false, "Unknown svd driver ", driver.value(), " in svd_cusolver computation.");
```
"torch.linalg.svd: ..."

mruberry commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -640,31 +718,72 @@ void svd_cusolver(const Tensor& A,
+    } else if (driver.value() == "gesvdaBatched") {
+      // gesvdaStridedBatched is preferred for "tall skinny" matrices
+      TORCH_CHECK(batch_size > 0 && m > 0 && n > 0 && m > n,
+        "Cusolver SVD `gesvdaBatched` driver requires the input matrices be non-empty "
```
Same note here -- nice explantations on both of these messages

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1533,6 +1549,10 @@
                                     `U` and `Vh`. Default: `True`.
 
 Keyword args:
+    driver (str, optional): name of the cuSOLVER method to be used. This keyword argument only works on CUDA inputs.
```
Nice documentation

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1480,6 +1480,22 @@
 
+.. note:: Selection of `driver` (this keyword argument only works on CUDA inputs with cuSOLVER backend):
+
+    - When `driver` is not given or set to `None` (*default*), a heuristic of drivers will be selected based on
```
"a heuristic of drivers will be selected matrix sizes and batch sizes of the inputs" -> " a driver will be selected automatically based on the shape of the inputs."

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1480,6 +1480,22 @@
+    - The `gesvdj` driver directly calls the cuSOLVER `cusolverDn<t>gesvdj` method. This Jacobi-based algorithm
+      has better performance than `gesvd` on GPU, but it may get less accurate results for ill-conditioned matrices or
+      matrices with very large sizes (be cautious if matrix size > 2048).
+    - The `gesvdjBatched` driver directly calls the cuSOLVER `cusolverDn<t>gesvdjBatched` method. It only works for
```
Remind me: were we thinking of the "gesvdj" calling either gesvdj or gesvdjBatched for the user?

xwang233 replied:
The discussion was to merge `gesvdj` and `gesvdjBatched` to one option, and select the actual driver in the backend based on input shapes.

Now I think it's a better idea to merge these two methods. With that, each of the three drivers would work for (nearly) all shapes.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1480,6 +1480,22 @@
+      matrices with very large sizes (be cautious if matrix size > 2048).
+    - The `gesvdjBatched` driver directly calls the cuSOLVER `cusolverDn<t>gesvdjBatched` method. It only works for
+      batched square matrices with size less than or equal to 32.
+    - The `gesvdaBatched` driver directly calls the cuSOLVER `cusolverDn<t>gesvdaStridedBatch` method. This is an
```
Is this driver only callable on batches? Maybe for user clarify we should rename it to `gesvda` if it can be called on two matrices or batches of matrices?

xwang233 replied:
`gesvda` driver will work on single matrix as well. I just removed the "Batched" in driver string name and changed it to `gesvda`.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1588,11 +1608,16 @@
 .. seealso::
 
         :func:`torch.linalg.svd` computes the full singular value decomposition.
+        Also check that documentation for the selection of `driver`.
```
Good pointer

Lezcano replied:
Nit. This might be unnecessary now as we are advertising this in the kwarg section?

mruberry reviewed (commented):
I'll let our linalg experts comment on the implementation. The docs and warning and errors look pretty reasonable -- I made a few small suggestions.

What are your thoughts on testing? Are there some special cases we should add to test_linalg.py that would demonstrate each driver doing what it should?

xwang233 committed:
cusolver gesvdj max sweeps in a constexpr

xwang233 committed:
reword error message

xwang233 committed:
language

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
merge gesvdj and gesvdjBatched drivers

xwang233 committed:
add svd driver tests

xwang233 commented:
Hi @mruberry , thanks for the review comments. I believe I've addressed most of them.

I also added new tests for those newly added svd drivers. The extra time cost on those tests is about 3 second on my machine.

xwang233 committed:
add the driver kwarg to test

xwang233 committed:
lint

xwang233 committed:
add gesvda test

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -2992,6 +2994,10 @@ TORCH_IMPL_FUNC(_linalg_svd_out)(const Tensor& A,
   // We need to distinguish the cuSOLVER case, as cuSOLVER expects F-contig matrices, but
   // it computes V rather than Vh
   const bool use_cusolver = at::native::svd_uses_cusolver(A);
+  if (!use_cusolver && driver.has_value()) {
```
Should this also check that `A` is on GPU?

Lezcano replied:
True that.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -640,31 +718,71 @@ void svd_cusolver(const Tensor& A,
+        convergence_check_needed = true;
+      }
+    } else if (driver.value() == "gesvda") {
+      // cuSOLVER: gesvdaStridedBatched is preferred for "tall skinny" matrices
+      TORCH_CHECK(batch_size > 0 && m > 0 && n > 0 && m > n,
+        "torch.linalg.svd: cuSOLVER `gesvda` driver requires the input matrices be non-empty "
+        "and number of rows be greater than number of columns (ideally tall and skinny).");
```
Empty matrices are handled at the beginning of the `_linalg_svd`  function, so the first 3 checks should not be necessary.

About the `m > n`  condition, we should be able to work around it as we do for `gesvd` by computing `A^H`. See
We should be able to work around this, as we do in `gesvd`, by transposing (and perhaps taking conjs as well) the input matrix.
https://github.com/pytorch/pytorch/blob/bf730e5039e8bf1614de6ad08ac6a24a1622dcbe/aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp#L438-L454

If this kernel does not support the `m == n` case (fair enough although a bit odd), this would make this torch check just report an error in that case.

xwang233 replied:
Thanks for the suggestions on empty input check!

I've once tried to transpose the input matrix in `gesvda`, like `gesvd` does, but it raised some cusolver errors. I'll try it again and either that works or I'll add some comments explaining this.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebra.cpp:
```diff
@@ -3111,7 +3122,7 @@ std::tuple<Tensor&, Tensor&, Tensor&> svd_out(const Tensor& self, bool some, boo
 }
 
-std::tuple<Tensor, Tensor, Tensor> svd(const Tensor& self, bool some, bool compute_uv) {
+std::tuple<Tensor, Tensor, Tensor> svd(const Tensor& self, bool some, bool compute_uv, c10::optional<c10::string_view> driver) {
```
I would suggest not to add a `driver` kwarg for this function, as it is deprecated and we will (hopefully) remove it soon-ish (cc @mruberry).
I would simply default this driver to be `gesvdj` and run with it.

Lezcano commented on aten/src/ATen/native/BatchLinearAlgebraKernel.cpp:
```diff
@@ -1018,6 +1018,7 @@ static void apply_svd(const Tensor& A,
 void svd_kernel(const Tensor& A,
                 const bool full_matrices,
                 const bool compute_uv,
+                c10::optional<c10::string_view> driver,
```
Add a `TORCH_INTERNAL_ASSERT` that this driver does not have a value, just in case.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -490,8 +491,8 @@ inline static void apply_svd_cusolver_gesvdj(const Tensor& A, const Tensor& U, c
   // Todo: expose the following two parameters to users
-  // TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetTolerance(gesvdj_params, 1.0e-7));
-  // TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetMaxSweeps(gesvdj_params, 15));
+  TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetTolerance(gesvdj_params, std::numeric_limits<scalar_t>::epsilon()));
+  TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjSetMaxSweeps(gesvdj_params, cusolver_gesvdj_max_sweeps));
```
This is great! Now, we should benchmark this just in case.
For example, benchmark it with random matrices, but also benchmark it with the matrices from the issue that started this discussion. For the matrices in that issue, we should also see if this change solves it or, at least, improves the state of things.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -515,13 +516,19 @@ inline static void apply_svd_cusolver_gesvdj(const Tensor& A, const Tensor& U, c
       gesvdj_params
     );
+
+    // The following code can be used to check or report the gesvdj residual.
+    // double residual = 0;
+    // TORCH_CUSOLVER_CHECK(cusolverDnXgesvdjGetResidual(handle, gesvdj_params, &residual));
+    // printf("gesvdj residual = %.6e\n", residual);
```
Thought: Should we check that the residual is less than some constant, and if it isn't then report that `gesvdj` was not able to converge and default to a different backend?

xwang233 replied:
`cusolverDnXgesvdjGetResidual` does introduce device-host sync which may negatively affect performance. Also, there isn't a good boundary value that can effectively distinguish if an ill-conditioned matrix is converged. In my experience, something like `4e-9` could be used, but not always effective.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -581,8 +588,73 @@ inline static void svd_cusolver_gesvdjBatched(const Tensor& A, const Tensor& U,
+    const Tensor& infos, bool full_matrices, bool compute_uv) {
+  using value_t = typename c10::scalar_value_type<scalar_t>::type;
+  int m = cuda_int_cast(A.size(-2), "m");
+  int n = cuda_int_cast(A.size(-1), "n");
```
Just to be sure, please add a `TORCH_INTERNAL_ASERT(m > n);`
Then, you can just use `n` in place of `k`, `S_stride` and `rank`.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -581,8 +588,73 @@ inline static void svd_cusolver_gesvdjBatched(const Tensor& A, const Tensor& U,
+  // cuSOLVER remark: if the user is confident on the accuracy of singular values and singular vectors,
+  //   for example, certain conditions hold (required singular value is far from zero),
+  //   then the performance can be improved by passing a null pointer to h_RnrmF, i.e. no computation of residual norm.
+  // Comment: although this residual is not checked or presented to end user, but it's still too dangerous to not calculate that.
```
If it's not used, why is it dangerous not to calculate it?

xwang233 replied:
I've checked with the cusolver team reworded the comment. We will pass `nullptr` to gesvda and not calculating that. This will improve the performance by a lot.

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -596,7 +668,12 @@ std::vector<int64_t> _check_gesvdj_convergence(const Tensor& infos, int64_t non_
     // From cusolver doc, if info < 0, the i-th function call parameter is wrong,
     // which means pytorch implementation of cusolver is wrong.
     TORCH_INTERNAL_ASSERT_DEBUG_ONLY(info_for_batch_i >= 0);
+
+    // In our use case, gesvdj, gesvdjBatched, and gesvdaStridedBatched have the same notations for `info`.
     if (info_for_batch_i == non_converging_info) res.push_back(i);
+
+    // However, it is not the same for gesvd, though we don't use this function to check gesvd convergence either.
+    // If it's implemented some day in the future, this needs to be handled carefully.
```
Fantastic comments! Thank you!

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -640,31 +718,71 @@ void svd_cusolver(const Tensor& A,
-  // Need to pass a copy of A, since A will be rewritten inside the function call
-  if (m <= 32 && n <= 32 && batch_size > 1 && (full_matrices || m == n)) {
-    svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+  const char* check_svd_doc = "Check doc at https://pytorch.org/docs/stable/generated/torch.linalg.svd.html";
```
Define this as a `static const char * const`

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -640,31 +718,71 @@ void svd_cusolver(const Tensor& A,
+        convergence_check_needed = true;
+      } else {
+        // gesvdj driver may be numerically unstable for large sized matrix
+        // Todo: Expose/adjust cusolver options of gesvdj number of iterations and numerical tolerance
```
Isn't this what you did above in this PR?

Lezcano commented on test/test_linalg.py:
```diff
@@ -2360,7 +2360,7 @@ def gen_error_message(input_size, p, keepdim, dim=None):
                 msg = gen_error_message(x.size(), p, keepdim)
                 self.assertEqual(res.shape, expected.shape, msg=msg)
-                self.assertEqual(res, expected, msg=msg, rtol=1.3e-6, atol=3e-4)
+                self.assertEqual(res, expected, msg=msg, rtol=4e-6, atol=6e-4)
```
How come we got a regression if we increased the number of iterations? If you don't know the reason I don't think you need to investigate, just curious.

xwang233 replied:
This is an example failing message on Windows only, does that give any hints? I honestly have no idea why it fails. There is no change to `norm`, and the `svd` default heuristic remains the same.

```
======================================================================
FAIL [0.016s]: test_norm_complex_old_cuda (__main__.TestLinalgCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 815, in dep_fn
    return fn(slf, *args, **kwargs)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 815, in dep_fn
    return fn(slf, *args, **kwargs)
  File "test_linalg.py", line 2363, in test_norm_complex_old
    self.assertEqual(res, expected, msg=msg, rtol=1.3e-6, atol=3e-4)
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 2170, in assertEqual
    assert_equal(
  File "C:\actions-runner\_work\pytorch\pytorch\build\win_tmp\build\torch\testing\_comparison.py", line 1074, in assert_equal
    raise error_metas[0].to_error()
AssertionError: complex norm failed for input size torch.Size([25, 25]), p=nuc, keepdim=True, dim=None
Tensor-likes are not close!

Mismatched elements: 1 / 1 (100.0%)
Greatest absolute difference: 0.0005035400390625 at index (0, 0) (up to 0.0003 allowed)
Greatest relative difference: 3.323382990382029e-06 at index (0, 0) (up to 1.3e-06 allowed)

```

Lezcano commented on test/test_linalg.py:
```diff
@@ -2584,34 +2584,48 @@ def test_svd(self, device, dtype):
+                    #   - cusolver 'gesvda' only works for m > n
+                    # - backend != 'cusolver' (driver should only be None)
+
+                    if driver == 'gesvda' and m <= n:
+                        continue
```
We can probably get rid of this condition, as mentioned above.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1588,11 +1610,16 @@
 Keyword args:
+    driver (str, optional): name of the cuSOLVER method to be used. This keyword argument only works on CUDA inputs.
+        If `None`, a default heuristic of algorithms will be used.
+        Available options are: `None`, `gesvd`, `gesvdj`, `gesvdjBatched`, and `gesvdaBatched`.
```
Remove batched options. 
Also, could you also point to the `linalg.svd` function docs here as well?

xwang233 replied:
oops, forgot svdvals

lezcano reviewed (commented):
Great stuff! It's so good to see that the previous refactor made PRs like this one so much easier to put together!

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
add internal_assert on svd_kernel on CPU

xwang233 committed:
address review comments

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
add transpose to gesvda

xwang233 committed:
fix comment

xwang233 committed:
gesvda doc

xwang233 committed:
gesvda: skip frobenius norm calculation

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
fix unused variable

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 commented:
Reviewers, can you help review this PR again? Thanks! cc @Lezcano @IvanYashchuk @mruberry @ngimel

Lezcano commented on aten/src/ATen/native/cuda/linalg/BatchLinearAlgebraLib.cpp:
```diff
@@ -812,31 +900,67 @@ void svd_cusolver(const Tensor& A,
-  if (!gesvdj_convergence_check.empty()) {
-    // fall back to gesvd path for those non-converging batches
+  if (!driver.has_value()) {
+    // use the default heuristics
+    if (gesvdj_batched_requirements) {
+      svd_cusolver_gesvdjBatched(cloneBatchedColumnMajor(A), U, S, V, info, compute_uv);
+      convergence_check_needed = true;
+    } else {
+      // gesvdj driver may be numerically unstable for large sized matrix
+      svd_cusolver_gesvdj(cloneBatchedColumnMajor(A), U, S, V, info, full_matrices, compute_uv);
+      convergence_check_needed = true;
+    }
```
The default heuristics are simply "driver == "gesvdj". It'd be simpler to do
```c++
// The default heuristic is to use gesvdj
auto driver_v = driver.value_or("gesvdj");
```
and get rid of this branch.

Lezcano commented on test/test_linalg.py:
```diff
@@ -2585,34 +2585,43 @@ def test_svd(self, device, dtype):
 
-            for batch, m, n in product(batches, ns, ns):
+            for batch, m, n, driver in product(batches, ns, ns, drivers):
+                if backend == 'cusolver' or driver is None:
+                    # only test cases below and skip otherwise:
+                    # - backend == 'cusolver' (driver can be anything)
+                    # - backend != 'cusolver' (driver should only be None)
+                    pass
+                else:
+                    continue
```
Perhaps simpler
```suggestion
                if not(backend == 'cusolver' or driver is None):
                    # only test cases below and skip otherwise:
                    # - backend == 'cusolver' (driver can be anything)
                    # - backend != 'cusolver' (driver should only be None)
                    continue
```

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1629,6 +1629,25 @@
+      but it may get less accurate results for ill-conditioned matrices or matrices with very large sizes
+      (be cautious if matrix size > 2048).
+    - The `gesvda` driver directly calls the cuSOLVER `cusolverDn<t>gesvdaStridedBatch` method. This is an
+      approximation method that is preferred when the number of rows is much different from the number of columns
```
```suggestion
      approximate method that is faster for very tall or very wide matrices.
```

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1682,6 +1701,10 @@
 
 Keyword args:
+    driver (str, optional): name of the cuSOLVER method to be used. This keyword argument only works on CUDA inputs.
+        If `None`, a default heuristic of algorithms will be used.
```
This line can probably go, as we've already dedicated a very long note to this.

lezcano reviewed (commented):
All of them are rather minor things. The only important thing is that `torch.svd` should not be modified as it's on the way to being deprecated. Once these are corrected this should be ready to go :)

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1629,6 +1629,25 @@
   Please use :func:`torch.linalg.svdvals`, which computes only the singular values,
   instead of `compute_uv=False`.
 
+.. note:: Selection of `driver` (this keyword argument only works on CUDA inputs with cuSOLVER backend):
+
+    - When `driver` is not given or set to `None` (*default*), a driver will be selected automatically
+      based on the shape of the inputs.
+      The default choice should be efficient and precise for most inputs. However, if input matrices
+      have large condition numbers or could be ill-conditioned, it's recommended to choose the `gesvd` driver
+      or use higher precision data type (e.g. use double instead of float).
+    - The `gesvd` driver directly calls the cuSOLVER `cusolverDn<t>gesvd` method. This QR-based algorithm is
+      less performant on GPU, but it can converge ill-conditioned matrices and get more precise results for those inputs
+      than the other drivers.
+    - The `gesvdj` driver calls the cuSOLVER `cusolverDn<t>gesvdjBatched` method if input batched matrices
+      are square matrices with sizes less than or equal to 32. Otherwise, the `gesvdj` driver calls the cuSOLVER
+      `cusolverDn<t>gesvdj` method. This Jacobi-based algorithm has better performance than `gesvd` on GPU,
+      but it may get less accurate results for ill-conditioned matrices or matrices with very large sizes
+      (be cautious if matrix size > 2048).
+    - The `gesvda` driver directly calls the cuSOLVER `cusolverDn<t>gesvdaStridedBatch` method. This is an
+      approximation method that is preferred when the number of rows is much different from the number of columns
+      (`m >> n` or `n >> m`).
+
```
I thought that, in the interest of conciseness and uniformity, it could be good to do as we do in `linalg.lstsq` and provide the user with a guide on how to choose the backend. What about rather than writing this note, we describe it as any other argument before the "differences with `np.linalg.svd`" section?
```
The `driver` kwarg may be used in CUDA with a cuSOLVER backend to choose the algorithm used to compute the SVD.
The choice of a driver is a trade-off between accuracy and speed.

- If :attr:`A` is well-conditioned (its `condition number`_ is not too large), or you do not mind some precision loss.

  - For a general matrix: `'gesvdj'` (Jacobi method)
  - If :attr:`A` is tall or wide (`m >> n` or `m << n`): `'gesvda'` (Approximate method)

- If :attr:`A` is not well-conditioned or precision is relevant: `'gesvd'` (QR based)

By default (:attr:`driver`\ `= None`), we call `'gesvdj'` and, if it fails, we fallback to `'gesvd'`.
```

xwang233 replied:
Done

xwang233 replied:
My local doc build is broken, gonna use the CI to check how the updated doc looks like.

xwang233 replied:
without the underscore after `condition number`

![image](https://user-images.githubusercontent.com/24860335/168976136-16d5e09f-d299-41ad-b082-a7c2e85ad866.png)

Lezcano replied:
Sorry, this is missing a link at the end of the docs of the form
```
.. _condition number:
    https://pytorch.org/docs/master/linalg.html#torch.linalg.cond
```
Check the docs for `linalg.lstsq`, as this is inspired by those.
Also, could you put ths, rather than in a note, in plain text before the section "differences with np.linalg.svd"? This is a description of an input to the function, so it should go in the main body of the docs.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
jit dispatch?

xwang233 committed:
review comments

xwang233 committed:
update doc

xwang233 committed:
use n instead of k in gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
lint

xwang233 committed:
fix a doc, underscore?

xwang233 committed:
update doc

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1626,6 +1626,15 @@
+The `driver` kwarg may be used in CUDA with a cuSOLVER backend to choose the algorithm used to compute the SVD.
+The choice of a driver is a trade-off between accuracy and speed.
+
+- If :attr:`A` is well-conditioned (its `condition number`_ is not too large), or you do not mind some precision loss.
```
This needs proper indentation in lines 2, 3 and 5 as otherwise:
https://docs-preview.pytorch.org/74521/generated/torch.linalg.svd.html

lezcano reviewed (approved):
Just a small detail in the docs, but otherwise this is ready to go!
Thank you for the great PR @xwang233 !

xwang233 committed:
change doc format

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
relax tolerance

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 commented:
Hi @mruberry @ngimel , can you help check if this PR is ready to go? Failures seem irrelevant.

Lezcano commented on torch/linalg/__init__.py:
```diff
@@ -1626,6 +1626,18 @@
 
 The parameter :attr:`full_matrices` chooses between the full (default) and reduced SVD.
 
+The `driver` kwarg may be used in CUDA with a cuSOLVER backend to choose the algorithm used to compute the SVD.
```
I realised after looking at the docs that this should be ``:attr:`driver` ``. Just add a fix for it if @IvanYashchuk has more comments about other things that need fixing.

IvanYashchuk reviewed (approved):
`apply_svd_cusolver_gesvdaStridedBatched` code looks good! Thank you for the comments explaining various choices made in the code to make cusolver not fail!
I like updated tolerance and max sweeps for the Jacobi-based algorithm.

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
Merge remote-tracking branch 'upstream/master' into cusolver-gesvda

xwang233 committed:
review comments

mruberry reviewed (approved):
Stamped!

Do we want to bother with benchmarks before merging? If not, let's just rebase this and once the tests pass use pytorchbot to merge it

xwang233 commented:
@pytorchbot merge this please
